## Obtaining the Efficient Frontier - Part II

*Suggested Answers follow (usually there are multiple ways to solve a problem in Python).*

Ok, let’s continue the exercise from the last lecture.

You already downloaded the data and generated two random weightings.

In [36]:
import numpy as np
import pandas as pd
import yfinance as yf
import matplotlib.pyplot as plt
%matplotlib inline

assets = ['WMT', 'META']
pf_data = pd.DataFrame()

for a in assets:
    pf_data[a] = yf.download(a, start='2014-01-01', auto_adjust=False)['Adj Close']


[*********************100%***********************]  1 of 1 completed
[*********************100%***********************]  1 of 1 completed


Mã lệnh khởi tạo bằng cách tính toán lợi suất logarit hàng ngày (log_returns) từ dữ liệu giá cổ phiếu WMT và META. Sau đó, nó tạo ra một bộ trọng số ngẫu nhiên đầu tiên

In [37]:
log_returns = np.log(pf_data / pf_data.shift(1))

num_assets = len(assets)

weights = np.random.random(num_assets)
weights /= np.sum(weights)
weights

array([0.71239022, 0.28760978])

Phân bổ vốn: Bộ trọng số này đại diện cho một danh mục đầu tư giả định với $\approx 71.24\%$ vốn vào WMT và $\approx 28.76\%$ vốn vào META

Now, estimate the expected Portfolio Return, Variance, and Volatility.

Expected Portfolio Return:

In [38]:
np.sum(weights * log_returns.mean()) * 250

np.float64(0.15396357811208633)

Expected Portfolio Variance:

In [39]:
np.dot(weights.T, np.dot(log_returns.cov() * 250, weights))

np.float64(0.040350830526002274)

Expected Portfolio Volatility:

In [40]:
np.sqrt(np.dot(weights.T,np.dot(log_returns.cov() * 250, weights)))

np.float64(0.2008751615456781)

***

The rest of this exercise will be a reproduction of what we did in the previous video.

thực hiện 1000 lần lặp để tạo ra hàng ngàn danh mục đầu tư ngẫu nhiên:

for x in range (1000)

Trong mỗi lần lặp, mã lệnh đều:Tạo ra bộ trọng số ngẫu nhiên mới (khác với lần trước).Tính toán Lợi suất hàng năm và Biến động hàng năm cho danh mục mới đó.Lưu trữ kết quả vào hai danh sách: pfolio_returns và pfolio_volatilities.

1)	Create two empty lists. Name them pf_returns and pf_volatilites.

In [41]:
pfolio_returns = []
pfolio_volatilities = []

2)	Create a loop with 1,000 iterations that will generate random weights, summing to 1, and will append the obtained values for the portfolio returns and the portfolio volatilities to pf_returns and pf_volatilities, respectively.

In [42]:
for x in range (1000):
    weights = np.random.random(num_assets)
    weights /= np.sum(weights)
    pfolio_returns.append(np.sum(weights * log_returns.mean()) * 250)
    pfolio_volatilities.append(np.sqrt(np.dot(weights.T,np.dot(log_returns.cov() * 250, weights))))

pfolio_returns, pfolio_volatilities

([np.float64(0.15018607167673906),
  np.float64(0.18409192582632034),
  np.float64(0.14604175458879748),
  np.float64(0.19903939281485014),
  np.float64(0.19990361683018137),
  np.float64(0.17862402733586696),
  np.float64(0.13461933990307356),
  np.float64(0.17823275073061282),
  np.float64(0.17673603233278978),
  np.float64(0.15967203118603218),
  np.float64(0.19941935994487575),
  np.float64(0.1922957014046599),
  np.float64(0.17942456777852078),
  np.float64(0.16233893843336059),
  np.float64(0.18574449118994824),
  np.float64(0.16287828022071715),
  np.float64(0.13850515336545688),
  np.float64(0.16090118294024114),
  np.float64(0.19191674350866147),
  np.float64(0.1868740120142375),
  np.float64(0.15142841057106216),
  np.float64(0.14940640564025248),
  np.float64(0.17189880320344583),
  np.float64(0.17505390794741013),
  np.float64(0.14319934231072898),
  np.float64(0.18906318515949588),
  np.float64(0.15500932093152328),
  np.float64(0.16588026859941635),
  np.float64(0.1637879

3)	Transform the obtained lists into NumPy arrays and reassign them to pf_returns and pf_volatilites. Once you have done that, the two objects will be NumPy arrays.

In [43]:
pfolio_returns = np.array(pfolio_returns)
pfolio_volatilities = np.array(pfolio_volatilities)

pfolio_returns, pfolio_volatilities

(array([0.15018607, 0.18409193, 0.14604175, 0.19903939, 0.19990362,
        0.17862403, 0.13461934, 0.17823275, 0.17673603, 0.15967203,
        0.19941936, 0.1922957 , 0.17942457, 0.16233894, 0.18574449,
        0.16287828, 0.13850515, 0.16090118, 0.19191674, 0.18687401,
        0.15142841, 0.14940641, 0.1718988 , 0.17505391, 0.14319934,
        0.18906319, 0.15500932, 0.16588027, 0.16378799, 0.15061202,
        0.15570515, 0.18712773, 0.16414519, 0.19366666, 0.16920725,
        0.16864765, 0.19823156, 0.17249039, 0.18836119, 0.14491198,
        0.16511863, 0.17500805, 0.15395735, 0.16431611, 0.19033868,
        0.19347876, 0.14454057, 0.17672649, 0.1815856 , 0.16601817,
        0.18044315, 0.16654294, 0.17899745, 0.17940329, 0.16596575,
        0.14105494, 0.14095937, 0.13536465, 0.18526145, 0.15303042,
        0.20322827, 0.16926721, 0.16759046, 0.16568015, 0.1388091 ,
        0.17367236, 0.16645164, 0.19689919, 0.15920374, 0.17278896,
        0.16679939, 0.13940199, 0.16942548, 0.19